optaplanner-core/src/main/java/org/optaplanner/core/impl/domain/variable/descriptor/BasicVariableDescriptor.java

```diff
@@ -0,0 +1,119 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.optaplanner.core.impl.domain.variable.descriptor;
+
+import org.optaplanner.core.api.domain.variable.PlanningVariable;
+import org.optaplanner.core.api.domain.variable.PlanningVariableGraphType;
+import org.optaplanner.core.impl.domain.common.accessor.MemberAccessor;
+import org.optaplanner.core.impl.domain.entity.descriptor.EntityDescriptor;
+import org.optaplanner.core.impl.domain.policy.DescriptorPolicy;
+
+public class BasicVariableDescriptor<Solution_> extends GenuineVariableDescriptor<Solution_> {
+
+    private boolean chained;
+    private boolean nullable;
+
+    // ************************************************************************
+    // Constructors and simple getters/setters
+    // ************************************************************************
+
+    public BasicVariableDescriptor(EntityDescriptor<Solution_> entityDescriptor, MemberAccessor variableMemberAccessor) {
+        super(entityDescriptor, variableMemberAccessor);
+    }
+
+    // ************************************************************************
+    // Lifecycle methods
+    // ************************************************************************
+
+    @Override
+    protected void processPropertyAnnotations(DescriptorPolicy descriptorPolicy) {
+        PlanningVariable planningVariableAnnotation = variableMemberAccessor.getAnnotation(PlanningVariable.class);
+        processNullable(descriptorPolicy, planningVariableAnnotation);
+        processChained(descriptorPolicy, planningVariableAnnotation);
+        processValueRangeRefs(descriptorPolicy, planningVariableAnnotation.valueRangeProviderRefs());
+        processStrength(descriptorPolicy, planningVariableAnnotation.strengthComparatorClass(),
+                planningVariableAnnotation.strengthWeightFactoryClass());
+    }
+
+    private void processNullable(DescriptorPolicy descriptorPolicy, PlanningVariable planningVariableAnnotation) {
+        nullable = planningVariableAnnotation.nullable();
+        if (nullable && variableMemberAccessor.getType().isPrimitive()) {
+            throw new IllegalArgumentException("The entityClass (" + entityDescriptor.getEntityClass()
+                    + ") has a @" + PlanningVariable.class.getSimpleName()
+                    + " annotated property (" + variableMemberAccessor.getName()
+                    + ") with nullable (" + nullable + "), which is not compatible with the primitive propertyType ("
+                    + variableMemberAccessor.getType() + ").");
+        }
+    }
+
+    private void processChained(DescriptorPolicy descriptorPolicy, PlanningVariable planningVariableAnnotation) {
+        chained = planningVariableAnnotation.graphType() == PlanningVariableGraphType.CHAINED;
+        if (chained && !variableMemberAccessor.getType().isAssignableFrom(entityDescriptor.getEntityClass())) {
```

Comment:
You make the right choice and reduce the complexity by extracting complex logic into a well-named boolean variable `chained`. Yet, for some reason, you choose not to do the same with the right hand side of the condition. Maybe you should have.

Reply:
I think I understand why you did that. You wanted to only execute the second check if the first one is true, saving some computation. Good thinking, but that problem can be avoided and, in my opinion, the end result is a method that reads even better:

    private void processChained(DescriptorPolicy descriptorPolicy, PlanningVariable planningVariableAnnotation) {
        chained = planningVariableAnnotation.graphType() == PlanningVariableGraphType.CHAINED;
        if (!chained) {
             return;
        }
        boolean isCorrectVariable = variableMemberAccessor.getType().isAssignableFrom(entityDescriptor.getEntityClass());
        if (!isCorrectVariable) {
            throw new IllegalArgumentException("The entityClass (" + entityDescriptor.getEntityClass()
                    + ") has a @" + PlanningVariable.class.getSimpleName()
                    + " annotated property (" + variableMemberAccessor.getName()
                    + ") with chained (" + chained + ") and propertyType (" + variableMemberAccessor.getType()
                    + ") which is not a superclass/interface of or the same as the entityClass ("
                    + entityDescriptor.getEntityClass() + ").\n"
                    + "If an entity's chained planning variable cannot point to another entity of the same class,"
                    + " then it is impossible to make chain longer than 1 entity and therefore chaining is useless.");
        }
        if (nullable) {
            throw new IllegalArgumentException("The entityClass (" + entityDescriptor.getEntityClass()
                    + ") has a @" + PlanningVariable.class.getSimpleName()
                    + " annotated property (" + variableMemberAccessor.getName()
                    + ") with chained (" + chained + "), which is not compatible with nullable (" + nullable + ").");
        }

Reply:
Nice improvement, indeed.

This part is copy-pasted from the `GenuineVariableDescriptor`, so it's an existing code that just moved from genuine to basic variable descriptor (because chaining is only applicable to basic variables).

I generally avoided refactoring existing code on the feature branch unless necessary to limit the amount of changes but this one is really small. I'm gonna do that, thanks!

Btw, your note made me look closer at genuine variable descriptor and I found more room for improvement there as well. But I'm going to keep that for `main`; it's really out of scope of this PR.